### PR TITLE
New 5GPM phasing pipeline

### PR DESCRIPTION
**Pull request by freerkvandijk**
* New pipeline fully tested
* Takes samplesheet containing 5GPM trio, specified:
  * DNA and RNA-seq data and sample IDs
  * Family relations
* Pipeline builds corresponding family trio structure plink files, collects RNA-seq data of parents and DNA-data from child. Afterwards does several QC Checks, including removal of known RNA-editting sites. Uses trio information to phase trio genotypes using Shapeit2.

### CODE REVIEW

File: compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+#Use bedtools to remove overlapping RNA-editting sites from VCF file, this results in unsorted file
+bedtools \
+intersect \
+-v \
+-a ${filteredFamilyVCF} \
+-b ${rnaEditBed} \
+-wa >> ${filteredRNAsitesFamilyVCF}.tmp.txt
+
+#Sort *.tmp.txt file by chromosome and position
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh]: why not pipe to sort? saves writing a file

File: compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+${stage} HTSlib/${htslibVersion}
+${checkStage}
+
+
+mkdir -p ${filteredFamilyVCFdir}
+
+
+#Use bedtools to remove overlapping RNA-editting sites from VCF file, this results in unsorted file
+bedtools \
+intersect \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh]: indent so it is easier to read

File: compute5/5GPM_phasing/protocols/ExtractSampleFromVCF.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+mkdir -p ${outputVCFdir}
+
+# Check if sample is not "none", if it is no data is available and no conversion analysis should be executed
+if [ "${sample}" != "none" ]
+then
+
+	# Extract only bi-allelic sites for samples of interest from VCF
+	java -Xmx8g -XX:ParallelGCThreads=2 -Djava.io.tmpdir=${TMPDIR} -jar ${EBROOTGATK}/GenomeAnalysisTK.jar \
+	-T SelectVariants \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/ExtractSampleFromVCF.sh]: indent makes it easier to read

**Comment by PatrickDeelen** [comment]
Hi, I think you should use --nomcmc in the shapeit step

File: compute5/5GPM_phasing/protocols/ConvertShapeitToVCF.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+${stage} shapeit/${shapeitVersion}
+${stage} HTSlib/${htslibVersion}
+${checkStage}
+
+
+#Run shapeit convert
+shapeit \
+-convert \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/ConvertShapeitToVCF.sh]: indent makes it easier to read

File: compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#MOLGENIS nodes=1 ppn=2 mem=8gb walltime=05:59:00
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/FilterRNAedittingSitesFromVCF.sh]: 2 ppn and 8gb necesarry?

File: compute5/5GPM_phasing/protocols/FilterVCF.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+${checkStage}
+
+mkdir -p ${filteredFamilyVCFdir}
+
+
+# Extract only bi-allelic sites for samples of interest from VCF
+# Also only select sites where all samples have a genotype
+# Do we still need to add other filtering options like GQ, DP, etc?
+java -Xmx8g -XX:ParallelGCThreads=2 -Djava.io.tmpdir=${TMPDIR} -jar ${EBROOTGATK}/GenomeAnalysisTK.jar \
+-T SelectVariants \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/FilterVCF.sh]: indent makes it easier to read

File: compute5/5GPM_phasing/protocols/MendelianErrorCheck.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+${stage} shapeit/${shapeitVersion}
+${checkStage}
+
+mkdir -p ${mendelianErrorCheckDir}
+
+#Run shapeit check to detect mendelian errors, which are written away in output log file
+shapeit \
+-check \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/MendelianErrorCheck.sh]: indent makes it easier to read

File: compute5/5GPM_phasing/protocols/MendelianErrorCheck.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
+#MOLGENIS nodes=1 ppn=2 mem=8gb walltime=05:59:00
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/MendelianErrorCheck.sh]: 2 ppn and 8gb necessary?

File: compute5/5GPM_phasing/protocols/MergeVCFs.sh [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#Extract mother RNA VCF from array by index
+MotherRNAVCFidx=$(get_index relations "mother")
+MotherRNAVCF=${outputSampleRNAVCF["$MotherRNAVCFidx"]}
+VCFs+=($MotherRNAVCF)
+
+toMerge=$(printf ' --variant %s ' $(printf '%s\n' ${VCFs[@]}))
+
+# Merge VCFs per family ID
+java -Xmx8g -XX:ParallelGCThreads=2 -Djava.io.tmpdir=${TMPDIR} -jar ${EBROOTGATK}/GenomeAnalysisTK.jar \
+-T CombineVariants \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/MergeVCFs.sh]: indent makes it easier to read

File: compute5/5GPM_phasing/protocols/PhaseFamily.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+
+${stage} shapeit/${shapeitVersion}
+${checkStage}
+
+mkdir -p ${phasedFamilyOutputDir}
+        
+#Run Shapeit2 for trio phasing
+shapeit \
+-B ${mendelianErrorCheckOutputPrefix} \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/PhaseFamily.sh]: indent

File: compute5/5GPM_phasing/protocols/PhaseFamily.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
+#MOLGENIS nodes=1 ppn=2 mem=8gb walltime=05:59:00
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/PhaseFamily.sh]: 2 ppn too many?

File: compute5/5GPM_phasing/protocols/RemoveMendelianErrors.sh [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	
+	#Create file containing IDs to remove
+	awk '{ if ($3 != "0") print $1}' FS="\t" ${mendelianErrorCheckOutputPrefix}.snp.me > ${mendelianErrorCheckOutputPrefix}.IDsToExclude.txt
+
+	#Remove first line of exclude file in place
+	sed -i '1,1d' ${mendelianErrorCheckOutputPrefix}.IDsToExclude.txt
+
+	#Run plink to exclude variants with mendelian errors
+	plink \
+	--bfile ${convertVCFtoPlinkPrefix} \
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/RemoveMendelianErrors.sh]: indent

File: compute5/5GPM_phasing/protocols/RemoveMendelianErrors.sh [hunk 13]
```diff
@@ -0,0 +1,71 @@
+#MOLGENIS nodes=1 ppn=2 mem=8gb walltime=05:59:00
```
Comment (npklein) [inline compute5/5GPM_phasing/protocols/RemoveMendelianErrors.sh]: doesnt need 2 ppn

**Comment by freerkvandijk** [comment]
I left the mcmc option specifically in, because it improves phasing of long
haplotypes. Besides this, on the testdata I've used the phasing itself only
takes minutes to complete so there is no need to apply this to speed up
analysis in my opinion.

**Comment by freerkvandijk** [comment]
Added indents and fixes

**Comment by PatrickDeelen** [comment]
If you have fewer than 10 samples you should not use mcmc. See 3B here: https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#reference

Instead you can use a reference panel as you are already doing to get the initial phasing. The duohmm will then make sure the phasing fits the trios.

**Comment by freerkvandijk** [comment]
Added --no-mcmc option to shapeit phasing

**Pull request merged**